flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/source/enumerator/splitter/MongoSampleSplitter.java

```diff
@@ -0,0 +1,140 @@
+ *       Defaults to: 64mb.
+ *   <li>scan.partition.samples: The number of samples to take per partition. Defaults to: 10. The
+ *       total number of samples taken is calculated as: {@code samples per partition * (count of
+ *       documents / number of documents per partition)}.
+ * </ul>
+ */
+@Internal
+public class MongoSampleSplitter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoSampleSplitter.class);
+
+    public static final MongoSampleSplitter INSTANCE = new MongoSampleSplitter();
+
+    private MongoSampleSplitter() {}
+
+    public Collection<MongoScanSourceSplit> split(MongoSplitContext splitContext) {
+        MongoReadOptions readOptions = splitContext.getReadOptions();
+        MongoNamespace namespace = splitContext.getMongoNamespace();
+
+        long count = splitContext.getCount();
+        long partitionSizeInBytes = readOptions.getPartitionSize().getBytes();
+        int samplesPerPartition = readOptions.getSamplesPerPartition();
+
+        long avgObjSizeInBytes = splitContext.getAvgObjSize();
+        if (avgObjSizeInBytes == 0L) {
+            LOG.info(
+                    "{} seems to be an empty collection, Returning a single partition.", namespace);
+            return MongoSingleSplitter.INSTANCE.split(splitContext);
+        }
+
+        long numDocumentsPerPartition = partitionSizeInBytes / avgObjSizeInBytes;
+        if (numDocumentsPerPartition >= count) {
+            LOG.info(
+                    "Fewer documents ({}) than the number of documents per partition ({}), Returning a single partition.",
+                    count,
+                    numDocumentsPerPartition);
+            return MongoSingleSplitter.INSTANCE.split(splitContext);
+        }
+
+        int numberOfSamples =
+                (int) Math.ceil((samplesPerPartition * count * 1.0d) / numDocumentsPerPartition);
+
+        List<BsonDocument> samples =
+                splitContext
+                        .getMongoCollection()
+                        .aggregate(
+                                Arrays.asList(
+                                        Aggregates.sample(numberOfSamples),
+                                        Aggregates.project(Projections.include(ID_FIELD)),
+                                        Aggregates.sort(Sorts.ascending(ID_FIELD))))
+                        .allowDiskUse(true)
+                        .into(new ArrayList<>());
+
+        List<MongoScanSourceSplit> sourceSplits = new ArrayList<>();
+        BsonDocument partitionStart = new BsonDocument(ID_FIELD, BSON_MIN_KEY);
+        int splitNum = 0;
+        for (int i = 0; i < samples.size(); i++) {
+            if (i % samplesPerPartition == 0 || i == samples.size() - 1) {
+                sourceSplits.add(
+                        createSplit(namespace, splitNum++, partitionStart, samples.get(i)));
+                partitionStart = samples.get(i);
+            }
+        }
```

Comment:
```suggestion
        BsonDocument partitionStart = samples.get(0);
        int splitNum = 0;
        for (int i = samplesPerPartition - 1; i < samples.size(); i+=samplesPerPartition) {
                sourceSplits.add(
                        createSplit(namespace, splitNum++, partitionStart, samples.get(i)));
                partitionStart = samples.get(i);
        }
```

Ideally this is how the loop would look like. We just have a list of points and split it into `numPartition` blocks. No need for special handling of MIN/MAX or to explicitly complete the right bound.